---
title: "Simulation Process and Needed Materials"
author: "Hillary Heiling"
date: "January 23, 2020"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Overview of Document

This document describes all of the necessary steps needed to simulate counts data for an IsoDeconvMM simulation, and states what information/materials would be needed for each step.

# Simulate Fragment Length Distribution Files

In the "order of operations", this step comes first.

1. Once all of the exon set counts have been simulated, sum up all counts to get the total counts for the sample, `total_cts`. 

2. Simulate `total_cts` lengths from a truncated normal distribution (mean 300, sd 50) truncated below at 150. 

3. Round all outputs from Step 2 to the nearest integer and tablulate the number of fragments simulated at each length. 

From this information, calculate the effective length design matrix.

# Find Exon Sets for all Isoforms

Materials needed:

* knownIsoforms .RData object relating the isoforms within each transcript cluster to the exons within the isoforms
* Information that connects each isoform to its gene ID. Possibility: nTE .RData object relates the cluster ID to its gene ID, but some clusters may have more than one gene associated with it. If so, additional informtaion may be needed.

Load objects of interest:

```{r}
# load the isoAll object:
load("Mouse_Materials/Mus_musculus.NCBIM37.67.nonoverlap.exon.knownIsoforms.RData")

# load the nTE object:
load("Mouse_Materials/Mus_musculus.NCBIM37.67.nTE.RData")

# load info matrix information from the final_geneMod object in "Step 1" output from IsoDeconvMM procedure
# Load the final_geneMod object from Step 1
load("Mouse_Materials/Step1_final_geneMod.RData")

```

Output of knownIsoforms and nTE:

```{r}
head(isoAll$chr1_1)

head(nTE)
```

In order to avoid problems mapping a particular gene ID to an exon within a cluster because there are multiple genes per cluster, we will filter the clusters to only consider those clusters containing a single gene. 

## What to do with information

Ideally want to compose names of exons sets in the format given to count files output from the countReads function.

```{r}
# Load example output of countReads 
counts_set1 = read.table(sprintf("Mouse_Materials/mm9_set1_counts.txt"), as.is = T)

head(counts_set1,10)
```

Want to create strings like the strings present in the second column:

clust_name|geneID|exon_number;

General procedure:

1. For each cluster, number all of the exons (in knownIsoforms matrix)

2. For each isoform, find which exons are present in the isoform (from the knownIsoforms matrix)

3. For each isoform, using the exon information from step 2, create string names for all exon sets with a single exon in the set: concatenate the cluster name, the gene ID (from nTE information and/or some other information) and exon number as described above. Store in a vector.

4. Using the single exon set information from Step 3, concatenate appropriate strings to create the double exon set information. Store in a vector.

Note: In order to determine which sets of exon combinations should be considered for possible exon sets, first need to determine effective lengths from fragment length distribution. 

5. Using the single exon set infomation from Step 3, concatenate appropriate strings to create the triple exon set information. Store in a vector.

Note:

Based on the exon set counts information from the mm9 simulated example, the counts are distributed as follows:

* Approximately 67.5% to 68.5% of total counts went to exon sets with a single exon in the set
* Approximately 27.5% to 28% of total counts went to exon sets with two exons in the set
* Approximately 3.5% of total counts went to exon sets with three exons in the set
* Approximately 0.5% of total counts went to exon sets with four or more exons in the set

Therefore, I think it is safe for our simulation just to consider single, double, and triple exon sets. 

# Simulate Negative Binomial Parameters for Exon Sets

## Total Reads to Gene Parameters

1. Specify total read counts for a sample (say, 20 million total reads)

2. Specify a Dirichlet distribution associated with all of the genes (say, 1000) within a sample. For one ``reference'' cell type, could just use a 'flat' Dirichlet distribution ($\alpha = 1$ for all genes). Alternatively, could model the Dirichlet to reflect what occurs in the UCLA data.

3. Sample a probability vector from the Dirichlet distribution in Step 2. 

4. Plug the probability vector from Step 3 into a multinomial distribution, and sample `n` = total read counts from this distribution. Do this `r` times, where `r` = number of reference samples per cell type. This will add additional biological variability between the pure cell type samples. Consider these count values to be the mean parameters for each gene. For each gene $g$, call the resulting fragment read count $T_g$.

5. Assign a constant dispersion parameter value across all genes. Specify the dispersion parameter such that the negative binomial variance is not much different than the variance for a poisson distribution. $Var(G) = \mu_g + \frac{1}{\theta}\mu_g^2$ with $\theta$ some large number, such as 100.

As a result of this step, we should have the equivalent of the negative binomial mean parameters for each gene cluster and a constant dispersion parameter for all genes.

## Gene Cluster to Isoform Parameters

1. For each gene, specify a Dirichlet distriubtion associated with the isoforms for the cluster

2. Sample a probability vector from the Dirichlet distribution in Step 1. Call this probability vector $\rho_g$ for gene $g$.

As a result of this step, we should have the probability vectors for the distribution of each isoform within a gene cluster.

## Isoform to Exon Set Parameters

### Available Materials and Known Models

From earlier, we have created the ExI effective length design matrix, $X_g = (x_{g1}, ..., x_{gI})$, for each gene cluster $g$. The columns of this design matrix correspond to the I isoforms for the gene cluster, and the rows correspond to the possible E exon sets for the gene cluster. We have also determined the negative binomial mean parameters for the gene cluster and each isoform within the gene cluster $g$. 

For each gene, let $\mu_g$ be a vector of means corresponding to all exon sets, $\mu_g = (\mu_{g1}, ..., \mu_{gE})^T$. This vector of means is characterized by $\mu_g = X_g \beta_g$ where $\beta_g = (\beta_{g1}, ..., \beta_{gI})^T$ and each $\beta_{gi}$ is proportional to the transcript abundance of isoform $i$. 

### Relationships:

Additional variables:

$\rho_g$ = probability vector of isoform proportion within a gene sampled from the Dirichlet dist in "Gene Cluster to Isoform Parameter" section

$r_g$ = scaling factor

$T_g$ = total fragment count for gene g (simulated in section "Total Reads to Gene Parameters")

$\mu_g = X_g\beta_g = X_g \rho_g r_g$, so $\beta_g = \rho_g r_g$. Need to find $r_g$

$T_g = \sum_e \mu_{ge} = r_g \sum_e (X_g \rho_g) \rightarrow r_g = \frac{T_g}{\sum_e (X_g \rho_g)}$

### Details to Simulate Exon Set Counts

1. If we simulate isoform probability distribution parameters as described in the "Gene Cluster to Isoform Parameters" section, we can calculate the $\beta_g$ parameters by calculating $\r_g = \frac{T_g}{\sum_e (X_g \rho_g)}$ and Then calculating $\beta_g = \rho_g r_g$.

2. Perform the matrix multiplication $X_g \beta_g$ to get the vector $\mu_g$. Each element of this vector will be the negative binomial mean parameter for each possible exon set. 

3. Using the negative binomial mean parameters from Step 2 and the constant $\theta$ parameter determined earlier, simulate exon set counts. 

### Potentially other useful information

Based on the exon set counts information from the mm9 simulated example, the counts are distributed as follows:

* Approximately 67.5% to 68.5% of total counts went to exon sets with a single exon in the set
* Approximately 27.5% to 28% of total counts went to exon sets with two exons in the set
* Approximately 3.5% of total counts went to exon sets with three exons in the set
* Approximately 0.5% of total counts went to exon sets with four or more exons in the set

# Multiple Cell Types to Consider

More specific details for this section are still needed. For now, the general idea:

Suppose, for now, that we assign a "reference" cell type, CT1, such that the dirichlet distributions for all of the relevant steps follow the UCLA data distributions for all of the relevant steps.  

Suppose that we also want to eventually have 3 cell types. We need to simulate negative binomial parameters for CT2 and CT3. 

Differences between cell types to consider:

* Differential isoform expression
* Differential isoform usage

Assume we choose 1000 genes. Of the 1000 genes, specify 100 to have differential isoform expression across cell types and 100 to have differential isoform usage across cell types

## Differential isoform expression

1. Create a fold change matrix of size NxJ, where N = number of genes (1000) and J = number of cell types. The column corresponding to the "reference" cell type 1 (CT1) will be, by default, all equal to 1. The other columns will have about 900 entries equal to 1 and the other 100 not equal to 1. (These factors are on the multiplicative scale). 

Note: The 100 entries != 1 should balance out in some way so that the total overall expression doesn't significantly change. In general terms, give some genes higher expression and some lower expression compared to the reference cell type. 

2. After the negative binomial mean parameters have been determined for each gene as described in the section "Total Reads to Gene Cluster Parameters", multiply these mean parameter values by the appropriate column of the NxJ fold change matrix. 

3. Intermediate step to balance total gene expression? (See below)

4. Proceed as typical for the rest of the process to get exon set counts as described above. 

### Ideas to balance the fold changes such that total expression is still similar:

Goal: To have the total proportion of gene expression for the 100 genes assigned differential expression to have the same proportion of gene expression relative to the total number of reads in both reference CT1 and new CTj. 

A. In reference cell type, calculate the gene expression proportion for the genes of interest by summing the negative binomial mean parameters for the genes and dividing by the total number of reads. 

B. Repeat Step A in the new cell type j (CTj).

C. Take the ratio of the proportions (proportion A / proportion B) and multiply this ratio by all of the negative binomial mean parameters in CTj. 

## Differential isoform usage

For the 100 genes designated to have a different isoform usage, perform the following:

In Step 1 of the section "Gene Cluster to Isoform Parameters", specify a different Dirichlet distribution to sample from. Otherwise, continue as normal. 

## Polyester package procedure (for inspiration, if needed)

This is what the polyester packae does (in general terms):

1. From the "reference" distribution, draw J counts from this distribution (J = total number of samples of interst) for each transcript.

2. Create a fold change matrix, N rows x J columns (rows = transcripts in the polyester package)

3. Multiply each count by the appropriate fold change value in the fold change matrix from Step 2

# Simulate Mixture Sample Counts

Could combine counts or sample from a mixture distribution ($p_1\mu_1 + ... + p_I\mu_i$)

Potential way to combine counts:

1. For each mixture sample, specify a probability vector (i.e. c(0.3, 0.3, 0.4), c(0.1, 0.5, 0.4) ...) where each element of the probability vector is associated with one of the pure cell types.

2. Multiply all counts (recall: all counts recorded by exon set) calculated for one of the pure cell type samples by the probability associated with that cell type. Do for all cell types. Round these new counts to the nearest integer.

3. Add all of the counts from step 2 together by exon set.

# UCLA Data Summary

Load summary data

```{r}
eDat = readRDS("Human_Materials/ucla_eDat.rds")
cDat = readRDS("Human_Materials/ucla_cDat.rds")
```

eDat exploration

```{r}
dim(eDat)
head(rownames(eDat),20)
colnames(eDat)
eDat[1:5, 1:5]
```

cDat exploration

```{r}

```

The End